ca/certificate-authority.go

```diff
@@ -418,14 +510,42 @@ func (ca *CertificateAuthorityImpl) IssueCertificate(csr x509.CertificateRequest
 
 		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
 		ca.log.Audit(fmt.Sprintf("PEM decode error, aborting and rolling back issuance: pem=[%s] err=[%v]", certPEM, err))
+		ca.SA.FinalizeCertificate(sr.cert)
 		tx.Rollback()
-		return emptyCert, err
+		return
 	}
-	certDER := block.Bytes
 
-	cert := core.Certificate{
-		DER:    certDER,
-		Status: core.StatusValid,
+	certObj, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		ca.log.Audit(fmt.Sprintf("Signer produced invalid DER, rolling back: serial=[%s] err=[%v]", serialHex, err))
+		ca.SA.FinalizeCertificate(sr.cert)
+		tx.Rollback()
+		return
+	}
+
+	// Fill in the remaining fields in the Certificate object
+	sr.cert.Status = core.StatusValid
+	sr.cert.Serial = core.SerialToString(certObj.SerialNumber)
+	sr.cert.Digest = core.Fingerprint256(certObj.Raw)
+	sr.cert.DER = certObj.Raw
+	sr.cert.Issued = time.Now()
+	sr.cert.Expires = certObj.NotAfter
+
+	// Fill in the remaining fields in the log event
+	logEvent.SerialNumber = sr.cert.Serial
+	logEvent.CommonName = certObj.Subject.CommonName
```

Comment:
Shouldn't we be logging the whole list of `hostNames` in addition to just the CN?